api-server/src/server/boot/challenge.js

```diff
@@ -328,6 +333,87 @@ function backendChallengeCompleted(req, res, next) {
     .subscribe(() => {}, next);
 }
 
+function createCoderoadChallengeCompleted(app) {
+  /* Example request coming from CodeRoad:
+   * req.body: { tutorialId: 'freeCodeCamp/learn-bash-by-building-a-boilerplate:v1.0.0' }
+   * req.headers: { coderoad-user-token: '8kFIlZiwMioY6hqqt...' }
+   */
+
+  const { WebhookToken, User } = app.models;
+
+  return async function coderoadChallengeCompleted(req, res) {
+    const { 'coderoad-user-token': userWebhookToken = null } = req.headers;
+    const { tutorialId = null } = req.body;
+
+    if (!tutorialId) return res.send(`'tutorialId' not found in request body`);
+
+    if (!userWebhookToken)
+      return res.send(`'coderoad-user-token' not found in request headers`);
+
+    const tutorialRepoPath = tutorialId?.split(':')[0];
+    const tutorialSplit = tutorialRepoPath?.split('/');
+    const tutorialOrg = tutorialSplit?.[0];
+    const tutorialRepoName = tutorialSplit?.[1];
+
+    // this allows any GH account to host the repo in development or staging
+    // .org submissions should always be from repos hosted on the fCC GH org
+    if (deploymentEnv !== 'staging' && environment !== 'development') {
+      if (tutorialOrg !== 'freeCodeCamp')
+        return res.send('Tutorial not hosted on freeCodeCamp GitHub account');
+    }
+
+    const codeRoadChallenges = getChallenges().filter(
+      challenge => challenge.challengeType === 12
+    );
+
+    // validate tutorial name is in codeRoadChallenges object
+    const tutorialInfo = codeRoadChallenges.find(tutorial =>
+      tutorial.url?.includes(tutorialRepoName)
+    );
+
+    if (!tutorialInfo) return res.send('Tutorial name is not valid');
+
+    const tutorialMongoId = tutorialInfo?.id;
+
+    try {
+      // check if webhook token is in database
+      const tokenInfo = await WebhookToken.findOne({
+        where: { id: userWebhookToken }
+      });
+
+      if (!tokenInfo) return res.send('User webhook token not found');
+
+      const { userId = null } = tokenInfo;
+
+      // check if user exists for webhook token
+      const user = await User.findOne({
+        where: { id: userId }
+      });
+
+      if (!user) return res.send('User for webhook token not found');
+
+      // submit challenge
+      const completedDate = Date.now();
+
+      const userUpdateInfo = buildUserUpdate(user, tutorialMongoId, {
+        id: tutorialMongoId,
+        completedDate
+      });
+
+      const updatedUser = await user.updateAttributes(
+        userUpdateInfo?.updateData
+      );
+
+      if (!updatedUser)
+        return res.send('An error occurred trying to submit the challenge');
+    } catch (e) {
+      return res.send('An error occurred trying to submit the challenge');
+    }
+
+    return res.send('Successfully submitted challenge');
```

Comment:
The request comes from coderoad, that's where the response goes. There's nothing they can do with that info. Basically, this route will just add it to our database. Their heatmap and completed challenges etc will be updated the next time they refresh the page. I didn't test it out - I'm hoping that when a user comes out of the VM, it will be updated - but, thinking about it, they will probably need a hard refresh.